anghammarad/src/main/scala/com/gu/anghammarad/Contacts.scala

```diff
@@ -2,34 +2,114 @@ package com.gu.anghammarad
 
 import com.gu.anghammarad.AnghammaradException.Fail
 import com.gu.anghammarad.models.{Contact, Email, EmailAddress, HangoutsChat, HangoutsRoom, Mapping, Message, _}
+import com.gu.anghammarad.Targets._
 
 import scala.util.{Success, Try}
 
 
 object Contacts {
   /**
     * Gets all available contacts for this target, from configuration.
+    *
+    * The logic is complex, the tests are a good reference for the expected behaviour.
+    *
+    * Exact matches are prioritised, then we apply logic to route other messages correctly.
+    * App, Stack and AWS Account use a hierarchy, App > Stack > AwsAccount.
+    * Stage must match - it is considered to be PROD if omitted.
     */
-  def resolveTargetContacts(targets: List[Target], mappings: List[Mapping]): Try[List[Contact]] = {
-    mappings.filter(_.targets == targets) match {
+  def resolveTargetContacts(rawTargets: List[Target], rawMappings: List[Mapping]): Try[List[Contact]] = {
+    val targets = normaliseStages(rawTargets)
+    val mappings = rawMappings.map(mapping => mapping.copy(targets = normaliseStages(mapping.targets)))
+
+    for {
+      exactMatches <- findExactMatches(targets, mappings)
+      underSpecifiedMatches = findUnderSpecifiedMatches(targets, mappings)
+      overSpecifiedMatches = findOverSpecifiedMatches(targets, mappings)
+      contacts <- exactMatches.orElse(underSpecifiedMatches).orElse(overSpecifiedMatches).fold[Try[List[Contact]]] {
+        Fail(s"Could not find matching contacts for $targets")
+      }(Success(_))
+    } yield contacts
+  }
+
+  /**
+    * Check for a mapping that exactly matches the target.
+    *
+    * Multiple exact matches is an error, so we fail straight away.
+    */
+  private def findExactMatches(targets: List[Target], mappings: List[Mapping]): Try[Option[List[Contact]]] = {
+    mappings.filter(_.targets.toSet == targets.toSet) match {
       case Nil =>
-        mappings.filter { case Mapping(mappingTargets, _) =>
-          targets.toSet subsetOf mappingTargets.toSet
-        } match {
-          case Nil =>
-            Fail(s"No contacts found for $targets")
-          case exactMatch :: Nil =>
-            Success(exactMatch.contacts)
-          case _ =>
-            Fail(s"Cannot resolve contacts from ambiguous partial matches for $targets")
-        }
+        Success(None)
       case exactMatch :: Nil =>
-        Success(exactMatch.contacts)
-      case _ =>
+        Success(Some(exactMatch.contacts))
+      case multipleMatches =>
         Fail(s"Found multiple exact matches while resolving contacts for $targets")
     }
   }
 
+  /**
+    * Searches among mappings that have more detail than requested.
+    *
+    * e.g.
+    * Ask for:
+    *   App("app")
+    * Mapping has:
+    *   Stack("stack"), App("app")
+    *
+    * This is tricky because we wouldn't to match on the following
```

Comment:
> This is tricky because we wouldn't to match on the following

Missing a word? 😄 

> This is tricky because we wouldn't **want** to match on the following

Reply:
Great spot, thank you for reading it so carefully :-)